lib/wallet/txdb.js

```diff
@@ -690,8 +690,18 @@ class TXDB {
 
       details.setOutput(i, path);
 
-      const credit = await this.getCredit(hash, i);
-      assert(credit);
+      let credit = await this.getCredit(hash, i);
+
+      if (!credit) {
+        // This credit didn't belong to us the first time we
+        // saw the transaction (before confirmation or rescan).
+        // Create new credit for database.
+        credit = Credit.fromTX(tx, i, height);
```

Comment:
Comparing the state changes that _would_ have happened if the credit was inserted via `insert()`. And most everything looks to be handled, with a few exceptions:
- It does not set `credit.own`, this may need to be set? 
- Since this tx existed before, the balance delta does not need to include the tx with `state.tx()`, correct?

Reply:
> Since this tx existed before, the balance delta does not need to include the tx with state.tx(), correct?

I think that's right, but could turn out to be another "bad assumption". As it is now, `state.tx()` is updated on `insert()`, called by `add()`. Once the tx is in the txdb (i.e. `this.getTX(hash)` is truthy) `add()` calls `confirm()` instead. If a never-before-seen tx is added _with_ a block, `insert()` handles the confirmation process. So I think it's safe to assume that if `confirm()` is called, the tx is already in the database and has already been counted in the balance.

> It does not set credit.own, this may need to be set?

Good catch. Looking into this, `credit.own` is only used in one place:

https://github.com/bcoin-org/bcoin/blob/3821b74a7e0d7335c6679098f39593957e8f080a/lib/wallet/wallet.js#L2043-L2049

Since we are in the process of confirming a coin, it's probably moot. But an extended edge case could potentially arise:

- Receive 2 output unconfirmed tx, we only know 1 address
- Second address is added to wallet
- tx is confirmed (new credit is created once this patch is merged)
- a chain reorg or wallet rescan causes this tx, and therefore both coins, to become _unconfirmed_
- wallet tries to spend the new credit added by this patch

In this double-edge case, the credit will _not_ be used by `wallet.getSmartCoins()`. This is not a terrible situation, but if the wallet sent the coin in the first place, we _could_ select this coin with confidence. This functionality could be added to this PR by adding back some of the logic present in `insert()`:

- If any of the tx _inputs_ were created by our own wallet (`getCredit()` and `getPath()` return truthy) set a local flag `own` to true, and add that property to the credit we create in the patch.

So we're talking about a situation where a user sends coins to an address they own but is not yet known by the wallet.

Waddaya think? Include it?